polaris/hub/storage.py

```diff
@@ -125,6 +129,101 @@ def _multipart_upload(self, key: str, value: bytes) -> None:
+    ) -> tuple[int, int, int]:
+        """
+        Copy the content of the source store to this store.
+
+        We leverage the internal knowledge of this store to make the operation more efficient than `zarr.copy_store`:
+            - Conditional `put_object` operation for the "skip" conflict resolution
+            - Parallel, concurrent `put_object` operations using a thread pool
+
+        Args:
+            source: source store
+            if_exists: behavior if the destination key already exists
+            log: optional logging function
+        """
+
+        def copy_key(key: str, source: Store, if_exists: ZarrConflictResolution) -> tuple[int, int, int]:
+            """
+            Sub-function, intended to be dispatched through a thread pool executor.
+            It will execute the copy operation for a single key from the source.
+            """
+            copied = skipped = bytes_copied = 0
+
+            with handle_s3_errors():
+                data = source[key]
+                if isinstance(data, memoryview):
+                    data = data.tobytes()
+
+                md5_hash = md5(data)
+                try:
+                    # The default is to replace the existing key.
+                    # Setting this header will raise an error if the key already exists.
+                    extra = {"IfNoneMatch": "*"} if if_exists == "skip" else {}
+                    self.s3_client.put_object(
+                        Bucket=self.bucket_name,
+                        Key=self._full_key(key),
+                        Body=data,
+                        ContentType=self.content_type,
+                        ContentMD5=b64encode(md5_hash.digest()).decode(),
+                        Metadata={
+                            "md5sum": md5_hash.hexdigest(),
+                        },
+                        **extra,
+                    )
+                    bytes_copied = buffer_size(data)
+                    copied = 1
+                except ClientError as error:
+                    error_code = error.response["Error"]["Code"]
+                    # Raised when the key already exists and IfNoneMatch is set
+                    if error_code == "PreconditionFailed":
+                        match if_exists:
+                            case "skip":
+                                skipped = 1
+                            case "raise":
+                                raise CopyError(f"key {key!r} exists in destination")
+                    else:
+                        raise
+            return copied, skipped, bytes_copied
+
+        source_store_version = getattr(source, "_store_version", 2)
+        if source_store_version != self._store_version:
+            raise ValueError("zarr stores must share the same protocol version")
+
+        # The executor will allocate min(nbr_CPU + 4, 32) threads
+        with ThreadPoolExecutor() as executor:
```

Comment:
**Question:** Don't want to scope creep too much here, but is there a simple way to add a progress bar of some sort? There is the "this may take a while" spinner, but that's annoyingly minimal for large datasets.

Reply:
I guess this can be done through the log function actually.

Reply:
The `log(f"Copied {total_copied} ({total_bytes_copied} bytes), skipped {total_skipped}, of {len(all_source_keys)} keys")` line from below gets printed every batch. I've passed in `logger.info` as the log function, and it shows up in the console as the upload takes place.

We could display a nicer message if we wanted, with a percentage of progress and some formatting to the bytes number.

Reply:
That works for now! Thanks!